readme.md
## Companion Drones

![](thumbnail.png)

--------------------------------------


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.0.25
Date: 2023-02-21
  Changes:
    - Fix desync related to refuel items.
---------------------------------------------------------------------------------------------------
Version: 1.0.24
Date: 2022-08-01
  Changes:
    - Fix problem with repairing.
---------------------------------------------------------------------------------------------------
Version: 1.0.23
Date: 2022-07-31
  Changes:
    - Fix crash with players and companions on other surfaces
    - Fixed not preserving item with tags data.
    - Maybe fix some cases of the robots getting stuck.
---------------------------------------------------------------------------------------------------
Version: 1.0.22
Date: 2021-12-02
  Changes:
    - QoL and reactivity improvements related to the job searching queue, submitted by madmaxoft.
    - Fix crash when loading with Alien biomes installed.
---------------------------------------------------------------------------------------------------
Version: 1.0.21
Date: 2021-02-22
  Changes:
    - Fix crash related to scenarios and players take 2.
---------------------------------------------------------------------------------------------------
Version: 1.0.20
Date: 2021-02-22
  Changes:
    - Fix crash related to scenarios and players.
---------------------------------------------------------------------------------------------------
Version: 1.0.19
Date: 2021-02-11
  Changes:
    - Made companions just spawn when the game starts instead of giving the items.
    - Fix companions becoming unresponsive when toggling combat.
---------------------------------------------------------------------------------------------------
Version: 1.0.18
Date: 2021-01-17
  Changes:
    - Fixed bumping into companions with Space exploration/Alien biomes installed.
---------------------------------------------------------------------------------------------------
Version: 1.0.17
Date: 2021-01-06
  Changes:
    - Fix crash using tile only blueprints.
    - Fixed the new shortcut not showing on old saves.
---------------------------------------------------------------------------------------------------
Version: 1.0.16
Date: 2021-01-05
  Changes:
    - Some fixes to the laser bot items.
    - Added shortcuts for enabling/disabling the combat and construction.
    - Made the companions more responsive to direct player deconstruction and blueprinting.
---------------------------------------------------------------------------------------------------
Version: 1.0.15
Date: 2020-12-12
  Changes:
    - Changing the way construction laser/robots are deployed to try to make them work nicer.
---------------------------------------------------------------------------------------------------
Version: 1.0.14
Date: 2020-12-08
  Changes:
    - Fix crash on entity destroyed in some circumstances.
    - Simple attempt to fix interaction with the SpaceExploration navigation view.
    - Make following a jetpacking player work a bit better.
---------------------------------------------------------------------------------------------------
Version: 1.0.13
Date: 2020-12-06
  Changes:
    - Prevent giving commands to the companions with spidertron remotes.
    - Fix weird case of drone being stuck following but not responding to stimuli.
---------------------------------------------------------------------------------------------------
Version: 1.0.12
Date: 2020-12-05
  Changes:
    - Micro things, like follow adjust, map icon, working with editor controller, etc.
---------------------------------------------------------------------------------------------------
Version: 1.0.11
Date: 2020-12-05
  Changes:
    - Fix unresponsive drones. Also added a user command just in case.
    - Added thumbnail.
---------------------------------------------------------------------------------------------------
Version: 1.0.10
Date: 2020-12-04
  Changes:
    - Fix for compatibility with Jetpack mod.
    - Fix for compatibility with Mining drones mod.
---------------------------------------------------------------------------------------------------
Version: 1.0.9
Date: 2020-12-04
  Changes:
    - Fix drones going inactive when they have items to give to the player.
    - Add check to make sure players don't start riding the companions.
---------------------------------------------------------------------------------------------------
Version: 1.0.8
Date: 2020-12-04
  Changes:
    - Removed the companion GUI.
    - Companions are minable in the normal way.
---------------------------------------------------------------------------------------------------
Version: 1.0.7
Date: 2020-12-04
  Changes:
    - I rewrote the updating logic, so drones that are just following the player will not take much script time.
    - Made the robots spawning and draggin behind the companions less of an issue.

data/beam_sprites.lua
local beam_blend_mode = "additive"
local beam_non_light_flags = { "trilinear-filtering" }
local path = "__Companion_Drones__/data/grey_beams/"

local function attach_beam_graphics(beam_table, blend_mode, beam_flags, beam_tint, light_tint)
  beam_table.start =
  {
    filename = path.."tileable-beam-START-wtf.png",
    flags = beam_flags or beam_non_light_flags,
    line_length = 4,
    width = 52,
    height = 40,
    frame_count = 16,
    direction_count = 1,
    shift = {-0.03125, 0},
    tint = beam_tint,
    hr_version =
    {
      filename = path.."hr-tileable-beam-START-wtf.png",
      flags = beam_flags or beam_non_light_flags,
      line_length = 4,
      width = 94,
      height = 66,
      frame_count = 16,
      direction_count = 1,
      shift = {0.53125, 0},
      tint = beam_tint,
      scale = 0.5
    }
  }

  beam_table.ending =
  {
    filename = path.."tileable-beam-END-wtf.png",
    flags = beam_flags or beam_non_light_flags,
    line_length = 4,
    width = 49,
    height = 54,
    frame_count = 16,
    direction_count = 1,
    shift = {-0.046875, 0},
    tint = beam_tint,
    hr_version =
    {
      filename = path.."hr-tileable-beam-END-wtf.png",
      flags = beam_flags or beam_non_light_flags,
      line_length = 4,
      width = 91,
      height = 93,
      frame_count = 16,
      direction_count = 1,
      shift = {-0.078125, -0.046875},
      tint = beam_tint,
      scale = 0.5
    }
  }

  beam_table.head =
  {
    filename = path.."beam-head-wtf.png",
    flags = beam_flags or beam_non_light_flags,
    line_length = 16,
    width = 45 - 7,
    height = 39,
    frame_count = 16,
    shift = util.by_pixel(-7/2, 0),
    tint = beam_tint,
    blend_mode = blend_mode or beam_blend_mode
  }

  beam_table.tail =
  {
    filename = path.."beam-tail-wtf.png",
    flags = beam_flags or beam_non_light_flags,
    line_length = 16,
    width = 45 - 6,
    height = 39,
    frame_count = 16,
    shift = util.by_pixel(6/2, 0),
    tint = beam_tint,
    blend_mode = blend_mode or beam_blend_mode
  }

  beam_table.body =
  {
    {
      filename = path.."beam-body-1-wtf.png",
      flags = beam_flags or beam_non_light_flags,
      line_length = 16,
      width = 32,
      height = 39,
      frame_count = 16,
      tint = beam_tint,
      blend_mode = blend_mode or beam_blend_mode
    },
    {
      filename = path.."beam-body-2-wtf.png",
      flags = beam_flags or beam_non_light_flags,
      line_length = 16,
      width = 32,
      height = 39,
      frame_count = 16,
      tint = beam_tint,
      blend_mode = blend_mode or beam_blend_mode
    },
    {
      filename = path.."beam-body-3-wtf.png",
      flags = beam_flags or beam_non_light_flags,
      line_length = 16,
      width = 32,
      height = 39,
      frame_count = 16,
      tint = beam_tint,
      blend_mode = blend_mode or beam_blend_mode
    },
    {
      filename = path.."beam-body-4-wtf.png",
      flags = beam_flags or beam_non_light_flags,
      line_length = 16,
      width = 32,
      height = 39,
      frame_count = 16,
      tint = beam_tint,
      blend_mode = blend_mode or beam_blend_mode
    },
    {
      filename = path.."beam-body-5-wtf.png",
      flags = beam_flags or beam_non_light_flags,
      line_length = 16,
      width = 32,
      height = 39,
      frame_count = 16,
      tint = beam_tint,
      blend_mode = blend_mode or beam_blend_mode
    },
    {
      filename = path.."beam-body-6-wtf.png",
      flags = beam_flags or beam_non_light_flags,
      line_length = 16,
      width = 32,
      height = 39,
      frame_count = 16,
      tint = beam_tint,
      blend_mode = blend_mode or beam_blend_mode
    }
  }

  beam_table.light_animations =
  {
    start =
    {
      filename = path.."hr-tileable-beam-START-light-wtf.png",
      line_length = 4,
      width = 94,
      height = 66,
      frame_count = 16,
      direction_count = 1,
      shift = {0.53125, 0},
      scale = 0.5,
      tint = light_tint
    },

    ending =
    {
      filename = path.."hr-tileable-beam-END-light-wtf.png",
      line_length = 4,
      width = 91,
      height = 93,
      frame_count = 16,
      direction_count = 1,
      shift = {-0.078125, -0.046875},
      scale = 0.5,
      tint = light_tint
    },

    head =
    {
      filename = path.."beam-head-light-wtf.png",
      line_length = 16,
      width = 45 - 7,
      height = 39,
      frame_count = 16,
      shift = util.by_pixel(-7/2, 0),
      tint = light_tint
    },

    tail =
    {
      filename = path.."beam-tail-light-wtf.png",
      line_length = 16,
      width = 45 - 6,
      height = 39,
      shift = util.by_pixel(6/2, 0),
      frame_count = 16,
      tint = light_tint
    },

    body =
    {
      {
        filename = path.."beam-body-1-light-wtf.png",
        line_length = 16,
        width = 32,
        height = 39,
        frame_count = 16,
        tint = light_tint
      },
      {
        filename = path.."beam-body-2-light-wtf.png",
        line_length = 16,
        width = 32,
        height = 39,
        frame_count = 16,
        tint = light_tint
      },
      {
        filename = path.."beam-body-3-light-wtf.png",
        line_length = 16,
        width = 32,
        height = 39,
        frame_count = 16,
        tint = light_tint
      },
      {
        filename = path.."beam-body-4-light-wtf.png",
        line_length = 16,
        width = 32,
        height = 39,
        frame_count = 16,
        tint = light_tint
      },
      {
        filename = path.."beam-body-5-light-wtf.png",
        line_length = 16,
        width = 32,
        height = 39,
        frame_count = 16,
        tint = light_tint
      },
      {
        filename = path.."beam-body-6-light-wtf.png",
        line_length = 16,
        width = 32,
        height = 39,
        frame_count = 16,
        tint = light_tint
      }
    }
  }

  return beam_table
end

return attach_beam_graphics

script/construction-laser.lua
local util = require("util")
local on_robot_built_entity = function(event)
  local robot = event.robot
  if not (robot and robot.valid) then return end
  if robot.name ~= "companion-construction-robot" then return end

  local entity = event.created_entity
  if not (entity and entity.valid) then return end

  local source = robot.logistic_network.cells[1].owner

  local duration = math.ceil((util.distance(source.position, entity.position) / 0.4) * 0.9)
  robot.surface.create_entity{name = "companion-build-beam", position = robot.position, target_position = entity.position, source = source, force = robot.force, source_offset = {0, -0}, duration = duration}

  if source.type == "spider-vehicle" then
    source.autopilot_destination =
    {
      (robot.position.x + source.position.x) / 2,
      (robot.position.y + source.position.y) / 2
    }
  end

end

local on_robot_mined_entity = function(event)
  local robot = event.robot
  if not (robot and robot.valid) then return end
  if robot.name ~= "companion-construction-robot" then return end

  local entity = event.entity
  if not (entity and entity.valid) then return end

  local source = robot.logistic_network.cells[1].owner
  local duration = math.ceil((util.distance(source.position, entity.position) / 0.4) / 2)
  robot.surface.create_entity{name = "companion-deconstruct-beam", position = robot.position, target_position = entity.position, source = source, force = robot.force, source_offset = {0, -0}, duration = duration}
  robot.surface.play_sound{position = robot.position, path = "utility/drop_item"}

  if source.type == "spider-vehicle" then
    source.autopilot_destination =
    {
      (robot.position.x + source.position.x) / 2,
      (robot.position.y + source.position.y) / 2
    }
  end

end

local spider_to_build = {}

local on_player_placed_equipment = function(event)
  local equipment = event.equipment
  if not (equipment and equipment.valid) then return end
  if equipment.name ~= "companion-roboport-equipment" then return end

  local player = game.get_player(event.player_index)
  if player.opened_gui_type ~= defines.gui_type.entity then return end

  local opened = player.opened
  if not (opened and opened.valid) then return end

  if true then
    spider_to_build[game.tick + 1] = opened
    game.print("HUKJE")
    return
  end
end

local on_tick = function()
  if not next(spider_to_build) then return end

  local opened = spider_to_build[game.tick]
  if not opened then return end
  spider_to_build[game.tick] = nil

  game.print("SHTSJH")
  local surface = opened.surface
  local position = opened.position
  local force = opened.force
  local network = opened.logistic_network

  if not network then error("WTF?") end
  for k = 1, 1 do
    local robot = surface.create_entity{name = "companion-construction-robot", position = {position.x + (math.random()-0.5), position.y + (math.random() - 0.5)}, force = force}
    robot.logistic_network = opened.logistic_network
    --surface.create_entity{name = "companion-beam", position = position, force = force, target = robot, source = opened}
  end

end

local lib = {}

lib.events =
{
  [defines.events.on_robot_built_entity] = on_robot_built_entity,
  [defines.events.on_robot_mined_entity] = on_robot_mined_entity,
  [defines.events.on_player_placed_equipment] = on_player_placed_equipment,
  [defines.events.on_tick] = on_tick
}

return lib

data.lua
local util = require("__core__/lualib/util")

local bot = util.copy(data.raw["construction-robot"]["construction-robot"])
bot.name = "companion-construction-robot"
bot.localised_name = "Companion construction laser bank. (DO NOT TOUCH)"
bot.max_payload_size = 5
bot.speed = 0.4
bot.max_speed = 0.4
bot.max_energy = "1000000MJ"
bot.energy_per_tick = "0J"
bot.speed_multiplier_when_out_of_energy = 1
bot.energy_per_move = "0J"
bot.min_to_charge = 0
bot.max_to_charge = 0
bot.working_sound = nil
bot.minable = {name = "fish", amount = 0, mining_time = 1}
bot.selection_box = {{-0.25,-0.25}, {0.25,0.25}}
bot.cargo_centered = {0, -1}
bot.selectable_in_game = false
bot.draw_cargo = true
bot.max_health = 9999999
bot.flags = {"hidden"}

bot.idle = util.empty_sprite()
bot.idle_with_cargo = util.empty_sprite()
bot.in_motion = util.empty_sprite()
bot.in_motion_with_cargo = util.empty_sprite()
bot.shadow_idle = util.empty_sprite()
bot.shadow_idle_with_cargo = util.empty_sprite()
bot.shadow_in_motion = util.empty_sprite()
bot.shadow_in_motion_with_cargo = util.empty_sprite()
bot.working = util.empty_sprite()
bot.shadow_working = util.empty_sprite()
bot.sparks = util.empty_sprite()
bot.smoke = nil
bot.water_reflection = nil
bot.placeable_by =
{
  {item = "companion-construction-robot", count = 1}
}
bot.created_effect =
{
  type = "direct",
  action_delivery =
  {
    type = "instant",
    target_effects =
    {
      {
        type = "script",
        effect_id = "companion-robot-spawned"
      }
    }
  }
}

local bot_item =
{
  type = "item",
  name = "companion-construction-robot",
  icon = "__Companion_Drones__/drone-icon.png",
  icon_size = 200,
  subgroup = "logistic-network",
  order = "a[robot]-b[construction-robot]",
  place_result = "companion-construction-robot",
  stack_size = 500,
  flags = {"hidden", "only-in-cursor"}
}

local equipment =
{
  type = "roboport-equipment",
  name = "companion-roboport-equipment",
  localised_name = {"companion-roboport"},
  take_result = "companion-roboport-equipment",
  sprite =
  {
    filename = "__base__/graphics/equipment/personal-roboport-equipment.png",
    width = 64,
    height = 64,
    priority = "medium"
  },

  shape =
  {
    width = 2,
    height = 2,
    type = "full"
  },

  energy_source =
  {
    type = "electric",
    buffer_capacity = "1MW",
    input_flow_limit = "1kW",
    usage_priority = "secondary-input",
  },

  charging_energy = "0kW",
  spawn_minimum = "0W",

  robot_limit = 6,
  construction_radius = 10,
  draw_construction_radius_visualization = false,
  spawn_and_station_height = 1,
  spawn_and_station_shadow_height_offset = 0,
  charge_approach_distance = 0,
  robots_shrink_when_entering_and_exiting = true,

  recharging_animation =
  {
    filename = "__base__/graphics/entity/roboport/roboport-recharging.png",
    priority = "high",
    width = 37,
    height = 35,
    frame_count = 16,
    scale = 1.5,
    animation_speed = 0.5
  },

  recharging_light = {intensity = 0.4, size = 5},
  stationing_offset = {0, -2},
  charging_station_shift = {0, -2},
  charging_station_count = 0,
  charging_distance = 0,
  charging_threshold_distance = 0,
  robot_vertical_acceleration = 10,
  categories = {"companion"}
}

local item_category =
{
  type = "item-subgroup",
  name = "companion",
  group = "combat",
  order = "dea-hank"
}

local equipment_item =
{
  type = "item",
  name = "companion-roboport-equipment",
  localised_name = {"companion-roboport"},
  icons =
  {
    {
      icon = "__Companion_Drones__/drone-icon.png",
      icon_size = 200
    },
    {
      icon = "__base__/graphics/equipment/personal-roboport-equipment.png",
      icon_size = 64,
      scale = 0.333,
    }
  },
  placed_as_equipment_result = "companion-roboport-equipment",
  subgroup = "companion",
  order = "c",
  default_request_amount = 1,
  stack_size = 20
}

local attach_beam_graphics = require("data/beam_sprites")

local build_beam = util.copy(data.raw["beam"]["electric-beam-no-sound"])
build_beam.name = "companion-build-beam"
build_beam.action = nil
attach_beam_graphics(build_beam, nil, nil, {0, 1, 0}, {0, 1, 0})

local deconstruct_beam = util.copy(data.raw["beam"]["electric-beam-no-sound"])
deconstruct_beam.name = "companion-deconstruct-beam"
deconstruct_beam.action = nil
attach_beam_graphics(deconstruct_beam, nil, nil, {1, 0, 0}, {1, 0, 0})

local inserter_beam = util.copy(util.copy(data.raw["beam"]["laser-beam"]))
inserter_beam.name = "inserter-beam"
--inserter_beam.head = util.empty_sprite()
--inserter_beam.head.repeat_count = 8
--inserter_beam.head =
--{
--  filename = "__Companion_Drones__/data//hr-fast-inserter-hand-closed.png",
--  priority = "extra-high",
--  width = 164,
--  height = 72,
--  scale = 0.25
--}
--inserter_beam.start = util.empty_sprite()
--inserter_beam.start.repeat_count = 8
--inserter_beam.body =
--{
--  filename = "__Companion_Drones__/data//hr-fast-inserter-hand-base.png",
--  priority = "extra-high",
--  height = 32,
--  width = 136,
--  scale = 0.25
--}
--inserter_beam.tail = util.empty_sprite()
--inserter_beam.tail.repeat_count = 8
--inserter_beam.ending = util.empty_sprite()
--inserter_beam.ending.repeat_count = 8
--inserter_beam.light_animations = nil
inserter_beam.target_offset = {0, 0}
inserter_beam.random_target_offset = false
inserter_beam.working_sound = nil
inserter_beam.damage_interval = 9999999
inserter_beam.action_triggered_automatically = false
inserter_beam.action = nil

local scale = 0.6
local leg_scale = 1
local arguments = {name = "spidertron"}
local drone =
{
  type = "spider-vehicle",
  name = "companion",
  localised_name = {"companion"},
  collision_box = {{-1 * scale, -1 * scale}, {1 * scale, 1 * scale}},
  selection_box = {{-1 * scale, -1 * scale}, {1 * scale, 1 * scale}},
  drawing_box = {{-3 * scale, -4 * scale}, {3 * scale, 2 * scale}},
  icon = "__Companion_Drones__/drone-icon.png",
  icon_size = 200,
  mined_sound = {filename = "__core__/sound/deconstruct-large.ogg",volume = 0.8},
  open_sound = { filename = "__base__/sound/spidertron/spidertron-door-open.ogg", volume= 0.35 },
  close_sound = { filename = "__base__/sound/spidertron/spidertron-door-close.ogg", volume = 0.4 },
  sound_minimum_speed = 0.3,
  sound_scaling_ratio = 0.1,
  allow_passengers = false,
  working_sound =
  {
    sound =
    {
      filename = "__base__/sound/spidertron/spidertron-vox.ogg",
      volume = 0.35
    },
    activate_sound =
    {
      filename = "__base__/sound/spidertron/spidertron-activate.ogg",
      volume = 0.5
    },
    deactivate_sound =
    {
      filename = "__base__/sound/spidertron/spidertron-deactivate.ogg",
      volume = 0.5
    },
    match_speed_to_activity = true
  },
  weight = 1,
  braking_force = 1,
  friction_force = 1,
  flags = {"placeable-neutral", "player-creation", "placeable-off-grid"},
  collision_mask = {},
  minable = {result = "companion", mining_time = 1},
  max_health = 250,
  resistances =
  {
    {
      type = "fire",
      decrease = 15,
      percent = 60
    },
    {
      type = "physical",
      decrease = 15,
      percent = 60
    },
    {
      type = "impact",
      decrease = 50,
      percent = 80
    },
    {
      type = "explosion",
      decrease = 20,
      percent = 75
    },
    {
      type = "acid",
      decrease = 0,
      percent = 70
    },
    {
      type = "laser",
      decrease = 0,
      percent = 70
    },
    {
      type = "electric",
      decrease = 0,
      percent = 70
    }
  },
  --corpse = "spidertron-remnants",
  --dying_explosion = "spidertron-explosion",
  energy_per_hit_point = 1,
  guns = {},
  inventory_size = 21,
  equipment_grid = "companion-equipment-grid",
  trash_inventory_size = 0,
  height = 2,
  torso_rotation_speed = 0.05,
  chunk_exploration_radius = 3,
  selection_priority = 45,
  graphics_set = spidertron_torso_graphics_set(0.6),
  base_render_layer = "smoke",
  render_layer = "air-object",
  energy_source =
  {
    type = "burner",
    fuel_category = "chemical",
    effectivity = 1,
    fuel_inventory_size = 3,
    smoke =
    {
      {
        name = "train-smoke",
        deviation = {0.3, 0.3},
        frequency = 100,
        position = {0, 0},
        starting_frame = 0,
        starting_frame_deviation = 60,
        height = 2,
        height_deviation = 0.5,
        starting_vertical_speed = -0.2,
        starting_vertical_speed_deviation = 0.1
      }
    }
  },
  movement_energy_consumption = "20kW",
  automatic_weapon_cycling = true,
  chain_shooting_cooldown_modifier = 0.5,
  spider_engine =
  {
    legs =
    {
      { -- 1
        leg = "companion-leg",
        mount_position = {0, -1},
        ground_position = {0, -1},
        blocking_legs = {1},
        leg_hit_the_ground_trigger = nil
      }
    },
    military_target = "spidertron-military-target"
  },

  minimap_representation =
  {
    filename = "__Companion_Drones__/drone-map.png",
    flags = {"icon"},
    size = {128, 128},
    scale = 0.25
  }

}
drone.graphics_set.render_layer = "air-entity-info-icon"
drone.graphics_set.base_render_layer = "air-object"
drone.graphics_set.autopilot_path_visualisation_line_width = 0
drone.graphics_set.autopilot_path_visualisation_on_map_line_width = 0
drone.graphics_set.autopilot_destination_visualisation = util.empty_sprite()
drone.graphics_set.autopilot_destination_queue_on_map_visualisation = util.empty_sprite()
drone.graphics_set.autopilot_destination_on_map_visualisation = util.empty_sprite()
drone.graphics_set.light =
{
  {
    type = "oriented",
    minimum_darkness = 0.3,
    picture =
    {
      filename = "__core__/graphics/light-cone.png",
      priority = "extra-high",
      flags = { "light" },
      scale = 1,
      width = 200,
      height = 200,
      shift = {0, -1}
    },
    source_orientation_offset = 0,
    shift = {0, (-200/32)- 0.5},
    add_perspective = false,
    size = 2,
    intensity = 0.6,
    color = {r = 0.92, g = 0.77, b = 0.3}
  }
}
drone.graphics_set.eye_light.size = 0

local leg =
{
  type = "spider-leg",
  name = "companion-leg",

  localised_name = {"entity-name.spidertron-leg"},
  collision_box = nil,
  collision_mask = {},
  selection_box = {{-0, -0}, {0, 0}},
  icon = "__base__/graphics/icons/spidertron.png",
  icon_size = 64, icon_mipmaps = 4,
  walking_sound_volume_modifier = 0,
  target_position_randomisation_distance = 0,
  minimal_step_size = 0,
  working_sound = nil,
  part_length = 1000000000,
  initial_movement_speed = 100,
  movement_acceleration = 100,
  max_health = 100,
  movement_based_position_selection_distance = 3,
  selectable_in_game = false,
  graphics_set = create_spidertron_leg_graphics_set(0, 1)
}

local layers = drone.graphics_set.base_animation.layers
for k, layer in pairs (layers) do
  layer.repeat_count = 8
  layer.hr_version.repeat_count = 8
end

table.insert(layers, 1,
{
  filename = "__base__/graphics/entity/rocket-silo/10-jet-flame.png",
  priority = "medium",
  blend_mode = "additive",
  draw_as_glow = true,
  width = 87,
  height = 128,
  frame_count = 8,
  line_length = 8,
  animation_speed = 0.5,
  scale = 1.13/4,
  shift = util.by_pixel(-0.5, 20),
  direction_count = 1,
  hr_version = {
    filename = "__base__/graphics/entity/rocket-silo/hr-10-jet-flame.png",
    priority = "medium",
    blend_mode = "additive",
    draw_as_glow = true,
    width = 172,
    height = 256,
    frame_count = 8,
    line_length = 8,
    animation_speed = 0.5,
    scale = 1.13/8,
    shift = util.by_pixel(-1, 20),
    direction_count = 1,
  }
})

local drone_item =
{
  type = "item-with-entity-data",
  name = "companion",
  icon = "__Companion_Drones__/drone-icon.png",
  icon_tintable = "__Companion_Drones__/drone-icon-tintable.png",
  icon_tintable_mask = "__Companion_Drones__/drone-icon-mask.png",
  icon_size = 200,
  subgroup = "companion",

  order = "a",
  stack_size = 1,
  place_result = "companion"
}

local gun =
{
  type = "active-defense-equipment",
  name = "companion-defense-equipment",
  localised_name = {"companion-laser"},
  sprite =
  {
    filename = "__base__/graphics/equipment/personal-laser-defense-equipment.png",
    width = 64,
    height = 64,
    priority = "medium"
  },
  shape =
  {
    width = 2,
    height = 2,
    type = "full"
  },
  energy_source =
  {
    type = "electric",
    usage_priority = "secondary-input",
    buffer_capacity = "1MJ"
  },

  attack_parameters =
  {
    type = "beam",
    warmup = 20,
    cooldown = 15,
    cooldown_deviation = 0.1,
    range = 21,
    --source_direction_count = 64,
    --source_offset = {0, -3.423489 / 4},
    damage_modifier = 1,
    ammo_type =
    {
      category = "laser",
      energy_consumption = "0.1KJ",
      action =
      {
        type = "direct",
        action_delivery =
        {
          type = "instant",
          target_effects =
          {
            type = "script",
            effect_id = "companion-attack"
          }
        }
      },
    }
  },

  automatic = true,
  categories = {"companion"}
}

local gun_item =
{
  type = "item",
  name = "companion-defense-equipment",
  localised_name = {"companion-laser"},
  icons =
  {
    {
      icon = "__Companion_Drones__/drone-icon.png",
      icon_size = 200
    },
    {
      icon = "__base__/graphics/equipment/personal-laser-defense-equipment.png",
      icon_size = 64,
      scale = 0.333,
    },
  },
  placed_as_equipment_result = "companion-defense-equipment",
  subgroup = "companion",
  order = "d",
  default_request_amount = 1,
  stack_size = 20
}

local plasma_projectile =
{
  type = "projectile",
  name = "companion-projectile",
  icon = "__Companion_Drones__/drone-icon.png",
  icon_size = 200,
  flags = {"not-on-map"},
  subgroup = "explosions",
  height = 1.4,
  rotatable = true,
  animation = nil,
  acceleration = 0.005,
  max_speed = 0.5,
  turn_speed = 0.001,
  turning_speed_increases_exponentially_with_projectile_speed = true,
  collision_box = {{-0.1, -0.1},{0.1, 0.1}},
  speed_modifier = {1, 0.707},
  hit_at_collision_position = true,
  force_condition = "enemy",
  action =
  {
    type = "direct",
    action_delivery =
    {
      type = "instant",
      target_effects =
      {
        {
          type = "create-entity",
          entity_name = "explosion"
        },
        {
          type = "damage",
          damage = {amount = 20, type = "laser"}
        },
      }
    }
  },
}

local companion_grid =
{
  type = "equipment-grid",
  name = "companion-equipment-grid",
  width = 10,
  height = 2,
  equipment_categories = {"companion"}
}


local shield =
{
  type = "energy-shield-equipment",
  name = "companion-shield-equipment",
  localised_name = {"companion-shield"},
  sprite =
  {
    filename = "__base__/graphics/equipment/energy-shield-equipment.png",
    width = 64,
    height = 64,
    priority = "medium"
  },
  shape =
  {
    width = 2,
    height = 2,
    type = "full"
  },
  max_shield_value = 250,
  energy_source =
  {
    type = "electric",
    buffer_capacity = "1MJ",
    input_flow_limit = "0.1MJ",
    usage_priority = "primary-input"
  },
  energy_per_shield = "0.01MJ",
  categories = {"companion"}
}

local shield_item =
{
  type = "item",
  name = "companion-shield-equipment",
  localised_name = {"companion-shield"},
  icons =
  {
    {
      icon = "__Companion_Drones__/drone-icon.png",
      icon_size = 200
    },
    {
      icon = "__base__/graphics/equipment/energy-shield-equipment.png",
      icon_size = 64,
      scale = 0.333,
    }
  },
  placed_as_equipment_result = "companion-shield-equipment",
  subgroup = "companion",
  order = "e",
  default_request_amount = 1,
  stack_size = 20
}

local battery =
{
  type = "battery-equipment",
  name = "companion-battery-equipment",
  sprite =
  {
    filename = "__base__/graphics/equipment/battery-equipment.png",
    width = 32,
    height = 64,
    priority = "medium"
  },
  shape =
  {
    width = 1,
    height = 2,
    type = "full"
  },
  energy_source =
  {
    type = "electric",
    buffer_capacity = "200000MJ",
    usage_priority = "tertiary"
  },
  categories = {"companion"}
}

local battery_item =
{
  type = "item",
  name = "companion-battery-equipment",
  icon = "__Companion_Drones__/drone-icon.png",
  icon_size = 200,
  placed_as_equipment_result = "companion-battery-equipment",
  subgroup = "companion",
  order = "e[robotics]-a[personal-roboport-equipment]",
  default_request_amount = 1,
  stack_size = 20
}

local category =
{
  type = "equipment-category",
  name = "companion"
}


local reactor =
{
  type = "generator-equipment",
  name = "companion-reactor-equipment",
  localised_name = {"companion-reactor"},
  sprite =
  {
    filename = "__base__/graphics/equipment/fusion-reactor-equipment.png",
    width = 128,
    height = 128,
    priority = "medium"
  },
  shape =
  {
    width = 2,
    height = 2,
    type = "full"
  },
  energy_source =
  {
    type = "electric",
    usage_priority = "primary-output"
  },
  power = "1MW",
  categories = {"companion"}
}

local reactor_item =
{
  type = "item",
  name = "companion-reactor-equipment",
  localised_name = {"companion-reactor"},
  icons =
  {
    {
      icon = "__Companion_Drones__/drone-icon.png",
      icon_size = 200
    },
    {
      icon = "__base__/graphics/equipment/fusion-reactor-equipment.png",
      icon_size = 128,
      scale = 0.333 * 0.5,
    }
  },
  placed_as_equipment_result = "companion-reactor-equipment",
  subgroup = "companion",
  order = "b",
  default_request_amount = 1,
  stack_size = 20
}

local recipes =
{
  {
    type = "recipe",
    name = "companion",
    enabled = true,
    energy_required = 15,
    ingredients =
    {
      {"electronic-circuit", 100},
      {"iron-gear-wheel", 50},
      {"iron-plate", 50},
      {"copper-cable", 50},
      {"raw-fish", 1},
    },
    result = "companion"
  },
  {
    type = "recipe",
    name = "companion-reactor-equipment",
    enabled = true,
    energy_required = 10,
    ingredients =
    {
      {"electronic-circuit", 100},
      {"iron-gear-wheel", 50}
    },
    result = "companion-reactor-equipment"
  },
  {
    type = "recipe",
    name = "companion-shield-equipment",
    enabled = true,
    energy_required = 10,
    ingredients =
    {
      {"iron-gear-wheel", 20},
      {"copper-plate", 20}
    },
    result = "companion-shield-equipment"
  },
  {
    type = "recipe",
    name = "companion-roboport-equipment",
    enabled = true,
    energy_required = 10,
    ingredients =
    {
      {"iron-plate", 20},
      {"iron-stick", 20}
    },
    result = "companion-roboport-equipment"
  },
  {
    type = "recipe",
    name = "companion-defense-equipment",
    enabled = true,
    energy_required = 10,
    ingredients =
    {
      {"iron-gear-wheel", 15},
      {"copper-cable", 10},
    },
    result = "companion-defense-equipment"
  }
}

data:extend(recipes)

local speed_sticker =
{
  type = "sticker",
  name = "speed-sticker",
  flags = {"not-on-map"},
  animation = util.empty_sprite(),
  duration_in_ticks = 100,
  target_movement_modifier_from = 1,
  target_movement_modifier_to = 1,
  vehicle_speed_modifier_from = 10,
  vehicle_speed_modifier_to = 1,
  vehicle_friction_modifier_from = 1,
  vehicle_friction_modifier_to = 1,
}

local attack_icon =
{
  filename = "__Companion_Drones__/drone-attack-shortcut.png",
  priority = "extra-high-no-scale",
  size = 200,
  scale = 1,
  flags = {"icon"},
}

local attack_icon_disabled =
{
  filename = "__Companion_Drones__/drone-attack-shortcut-disabled.png",
  priority = "extra-high-no-scale",
  size = 200,
  scale = 1,
  flags = {"icon"},
}

local attack_shortcut =
{
  type = "shortcut",
  name = "companion-attack-toggle",
  localised_name = {"companion-attack-toggle"},
  order = "a[companion-drones]",
  action = "lua",
  style = "default",
  icon = attack_icon,
  small_icon = attack_icon,
  disabled_icon = attack_icon_disabled,
  disabled_small_icon = attack_icon_disabled,
  toggleable = true
}


local construct_icon =
{
  filename = "__Companion_Drones__/drone-construction-shortcut.png",
  priority = "extra-high-no-scale",
  size = 200,
  scale = 1,
  flags = {"icon"},
}

local construct_icon_disabled =
{
  filename = "__Companion_Drones__/drone-construction-shortcut-disabled.png",
  priority = "extra-high-no-scale",
  size = 200,
  scale = 1,
  flags = {"icon"},
}

local construct_shortcut =
{
  type = "shortcut",
  name = "companion-construction-toggle",
  localised_name = {"companion-construction-toggle"},
  order = "a[companion-drones]",
  action = "lua",
  style = "default",
  icon = construct_icon,
  small_icon = construct_icon,
  disabled_icon = construct_icon_disabled,
  disabled_small_icon = construct_icon_disabled,
  toggleable = true
}

data:extend
{
  bot,
  bot_item,
  item_category,
  equipment,
  equipment_item,
  --build_beam,
  --deconstruct_beam,
  inserter_beam,
  drone,
  drone_item,
  leg,
  gun,
  gun_item,
  plasma_projectile,
  companion_grid,
  shield,
  shield_item,
  --battery,
  --battery_item,
  reactor,
  reactor_item,
  category,
  speed_sticker,
  attack_shortcut,
  construct_shortcut

}
